### PR TITLE
'Undefined property' in case of a wrong auth code payload

### PR DESCRIPTION
**Pull request by x0st**
I encountered the error `Undefined property: stdClass::$auth_code_id` when passed a refresh token to the field `code` instead of a real auth code.

image

My code checks if the auth code's payload is correct: all the necessary fields are checked for presence.

### CODE REVIEW

File: src/Grant/AuthCodeGrant.php [hunk 1]
```diff
@@ -178,6 +181,29 @@ public function respondToAccessTokenRequest(
...
+            isset($authCodePayload->client_id),
+            isset($authCodePayload->redirect_uri),
+            isset($authCodePayload->scopes) && is_array($authCodePayload->scopes),
+            isset($authCodePayload->user_id)
+        ];
+
+        if (in_array(false, $errors, true)) {
+            throw OAuthServerException::invalidRequest('code');
+        }
+    }
```
Comment (sergeyklay) [inline src/Grant/AuthCodeGrant.php]: Looks like a pasta code here.

1. As you may know `json_decode` may return NULL  if the cannot be decoded or if the encoded data is deeper than the recursion limit. So, why do we need type hinting here?

2. I think the code may be more simple and obviously faster:

```php
$validCodePayload = 
            isset($authCodePayload->expire_time) &
            isset($authCodePayload->auth_code_id) &
            isset($authCodePayload->client_id) &
            isset($authCodePayload->redirect_uri) &
            isset($authCodePayload->scopes) &&
            is_array($authCodePayload->scopes) &
            isset($authCodePayload->user_id);

if ($validCodePayload === false) {
    throw OAuthServerException::invalidRequest('code');
}
```

3. I wold like to get hint - what exactly is missing

4. IMO is it better to validate JSON by using JSON Schema

Comment (x0st) [reply src/Grant/AuthCodeGrant.php]: Thanks for reviewing. I will fix it ASAP

Comment (sergeyklay) [reply src/Grant/AuthCodeGrant.php]: I think that 3 and 4 aren't mandatory, although for me they are still important.

**Comment by Sephster** [comment]
Thanks for this PR. Sorry that you hit this issue. I agree, the system should probably be checking for this property (and others) before using it. I have a couple of thoughts and observations that might help:

Have you considered using JSON Schema for this? I think it might be a good fit for validating the format of the payload and the [League JSON guard package](https://github.com/thephpleague/json-guard) would be suitable for this purpose.

I think using a schema to validate against would make this a cleaner and clearer implementation if it can be achieved so it would be good to explore the feasibility of this if not done already. Would you be willing to take a look at this? Thanks again for your PR

**Comment by x0st** [comment]
@Sephster What do you think about [justinrainbow/json-schema](https://github.com/justinrainbow/json-schema)? I think it is easier to use and more understandable.

**Pull request closed without merging**